hibernate-core/src/main/java/org/hibernate/dialect/aggregate/OracleAggregateSupport.java

```diff
@@ -138,27 +145,69 @@ public String aggregateComponentCustomReadExpression(
 										placeholder,
 										"json_value(" + parentPartExpression + columnExpression + "' returning " + column.getColumnDefinition() + ')'
 								);
+
 							case DATE:
-								return template.replace(
-										placeholder,
-										"to_date(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD')"
-								);
+								if (this.dateTypesStoreAsString) {
+									return template.replace(
+											placeholder,
+											"to_date(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD')"
+									);
+								}
+								else {
+									// Oracle OSON extension is used, value is not stored as string
+									return template.replace(
+											placeholder,
+											"json_value(" + parentPartExpression + columnExpression + "' returning date)"
+									);
+								}
+
 							case TIME:
 								return template.replace(
 										placeholder,
 										"to_timestamp(json_value(" + parentPartExpression + columnExpression + "'),'hh24:mi:ss')"
 								);
 							case TIMESTAMP:
-								return template.replace(
-										placeholder,
-										"to_timestamp(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD\"T\"hh24:mi:ss.FF9')"
-								);
+								if (this.dateTypesStoreAsString) {
+									return template.replace(
+											placeholder,
+											"to_timestamp(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD\"T\"hh24:mi:ss.FF9')"
+									);
+								}
+								else {
+
+									return template.replace(
+											placeholder,
+											"json_value(" + parentPartExpression + columnExpression + "' returning timestamp)"
+									);
+								}
+							case DURATION:
+								if (this.dateTypesStoreAsString) {
+									return template.replace(
+											placeholder,
+											"cast(json_value(" + parentPartExpression + columnExpression + "') as " + column.getColumnDefinition() + ')'
+									);
+								}
+								else {
+									return template.replace(
+											placeholder,
+											"json_value(" + parentPartExpression + columnExpression + "' returning interval day to second)"
+									);
+								}
 							case TIMESTAMP_WITH_TIMEZONE:
 							case TIMESTAMP_UTC:
-								return template.replace(
-										placeholder,
-										"to_timestamp_tz(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD\"T\"hh24:mi:ss.FF9TZH:TZM')"
-								);
+								if (this.dateTypesStoreAsString) {
+									return template.replace(
+											placeholder,
+											"to_timestamp_tz(json_value(" + parentPartExpression + columnExpression + "'),'YYYY-MM-DD\"T\"hh24:mi:ss.FF9TZH:TZM')"
+									);
+								}
+								else {
+									// Oracle OSON extension is used, value is not stored as string
+									return template.replace(
+											placeholder,
+											"json_value(" + parentPartExpression + columnExpression + "')"
```

Comment:
```suggestion
											"json_value(" + parentPartExpression + columnExpression + "' returning timestamptz(9))"
```

Looks like something is missing.